File: client/src/main/java/au/com/codeka/warworlds/client/opengl/TextureManager.java
package au.com.codeka.warworlds.client.opengl;

import android.content.Context;

import androidx.annotation.Nullable;

import com.google.common.base.Preconditions;

import java.lang.ref.WeakReference;
import java.util.HashMap;

/**
 * {@link TextureManager} manages {@link BitmapTexture}s that we have loaded and ensures we never
 * load the same texture more than once.
 */
public class TextureManager {
  private final Context context;
  private final HashMap<String, WeakReference<BitmapTexture>> cache = new HashMap<>();

  public TextureManager(Context context) {
    this.context = Preconditions.checkNotNull(context);
  }

  /** Loads the texture with the given name, or returns null if the texture could not be loaded. */
  @Nullable
  public BitmapTexture loadTexture(String name) {
    WeakReference<BitmapTexture> ref = cache.get(name);
    if (ref != null) {
      BitmapTexture bmp = ref.get();
      if (bmp != null) {
        return bmp;
      }
    }

    BitmapTexture bmp = BitmapTexture.load(context, name);
    cache.put(name, new WeakReference<>(bmp));
    return bmp;
  }

  /** Loads a texture from the given URL. */
  public BitmapTexture loadTextureUrl(String url) {
    WeakReference<BitmapTexture> ref = cache.get(url);
    if (ref != null) {
      BitmapTexture bmp = ref.get();
      if (bmp != null) {
        return bmp;
      }
    }

    BitmapTexture bmp = BitmapTexture.loadUrl(context, url);
    cache.put(url, new WeakReference<>(bmp));
    return bmp;
  }
}

File: client/src/main/kotlin/au/com/codeka/warworlds/client/opengl/TextureManager.kt
package au.com.codeka.warworlds.client.opengl

import android.content.Context
import au.com.codeka.warworlds.client.opengl.BitmapTexture.Companion.load
import au.com.codeka.warworlds.client.opengl.BitmapTexture.Companion.loadUrl
import com.google.common.base.Preconditions
import java.lang.ref.WeakReference
import java.util.*

/**
 * [TextureManager] manages [BitmapTexture]s that we have loaded and ensures we never
 * load the same texture more than once.
 */
class TextureManager(private val context: Context) {
  private val cache = HashMap<String, WeakReference<BitmapTexture>>()

  /** Loads the texture with the given name, or returns null if the texture could not be loaded.  */
  fun loadTexture(name: String): BitmapTexture? {
    synchronized(cache) {
      val ref = cache[name]
      if (ref != null) {
        val bmp = ref.get()
        if (bmp != null) {
          return bmp
        }
      }
      val bmp = load(context, name)
      cache[name] = WeakReference(bmp)
      return bmp
    }
  }

  /** Loads a texture from the given URL.  */
  fun loadTextureUrl(url: String): BitmapTexture {
    synchronized(cache) {
      val ref = cache[url]
      if (ref != null) {
        val bmp = ref.get()
        if (bmp != null) {
          return bmp
        }
      }
      val bmp = loadUrl(context, url)
      cache[url] = WeakReference(bmp)
      return bmp
    }
  }
}
